### PR TITLE
[RFC only] Shared memory support definitions

### PR DESCRIPTION
**Pull request by shefty**
This series defines a set of data structures and helper routines that will be used to implement a set of shared memory primitives and provider.

1. Modify the cirque template to allow the user to allocate the memory region, rather than the template allocating the data buffers.
2. Introduce a new template that provides a fixed sized buffer pool, which also allows users to allocate the memory that is used.
3. Define the shm data structures that will be exposed between different processes.
4. Define the protocol structures for implementing libfabric features over shared memory.

Ideally, the shared memory protocol will have some overlap with the protocol used by the utility provider(s).

### CODE REVIEW

**Comment by shefty** [comment]
@ddurnov - please review the shm data structures.  I hope to build out from them.  For scalability the shm provider will use a single receive cmd queue, protected by an atomic, rather than one per peer.  Though the primitives could be adapted for either.  Large data transfers will use process_vm_readv/process_vm_writev, versus bounce buffering.  This will gives us additional bandwidth performance with lower memory/CPU/power utilitization, and a simpler implementation.

@jithinjosepkl - please review the freestack template.  I think it could be useful as a base data structure for implementing a tagged memory or fi_context free pool.  It's a derivation of the util_buf_pool adapted for placement into a shared memory segment.

File: include/fi_mem.h [hunk 1]
```diff
@@ -62,15 +62,69 @@ static inline void *mem_dup(const void *src, size_t size)
 	return dest;
 }
 
+
+/*
+ * Buffer pool (free stack) template
+ */
+#define FREESTACK_EMPTY	-1
```
Comment (bturrubiates) [inline include/fi_mem.h]: Should this be `FREESTACK_EMPTY -1`?

Comment (shefty) [inline include/fi_mem.h]: > > +#define FREESTACK_EMPTY	-1
> 
> Should this be FREESTACK_EMPTY -1?

I'm missing the difference here

File: include/fi_mem.h [hunk 2]
```diff
@@ -62,15 +62,69 @@ static inline void *mem_dup(const void *src, size_t size)
...
+
+#define freestack_isempty(fs)	((fs)->next == FREESTACK_EMPTY)
+#define freestack_push(fs, x)					\
+{								\
+	*(int *) &(fs)->buf[x] = (fs)->next;			\
```
Comment (bturrubiates) [inline include/fi_mem.h]: Why casting to an `int *` here? 

Shouldn't this be part of the freestack template and be casting to `entrytype`?

File: include/fi_mem.h (hunk 1 above)
Comment (bturrubiates) [inline include/fi_mem.h]: Oh strange, the Github diff viewer is missing the space but if I look at the raw version it has the space...

File: include/fi_shm.h [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	int		pid;
+	atomic_t	lock;
+	int		peer_size;
+
+	size_t		total_size;
+	size_t		rx_cmd_offset;
+	size_t		rx_ctrl_offset;
+	size_t		inject_buf_offset;
+
+	struct shm_region *peer[1];
```
Comment (goodell) [inline include/fi_shm.h]: I take it this is intended to be an array with variable length?  If so, I think it's a lot cleaner to write it as [a proper flexible array member](https://en.wikipedia.org/wiki/Flexible_array_member) (with empty `[]`).

File: include/fi_mem.h [hunk 4]
```diff
@@ -62,15 +62,69 @@ static inline void *mem_dup(const void *src, size_t size)
...
+	assert(!freestack_isempty(fs));				\
+	(fs)->next = (int) (fs)->buf[i];			\
+	return i;						\
+}
+
+#define DECLARE_FREESTACK(entrytype, name)			\
+struct name {							\
+	size_t		size;					\
+	size_t		next;					\
+	entrytype	buf[1];					\
```
Comment (goodell) [inline include/fi_mem.h]: ditto here on `[]` vs `[1]`

File: include/fi_shm.h [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	char		*name;
+	size_t		cmd_size;
+	size_t		cmd_count;
+	size_t		ctrl_size;
+	size_t		ctrl_count;
+	size_t		tx_size;
+	size_t		tx_count;
+	size_t		inject_size;
+	size_t		inject_count;
+	size_t		peer_count;
```
Comment (hppritcha) [inline include/fi_shm.h]: What is the intent of the ```peer_count``` field?

Comment (shefty) [inline include/fi_shm.h]: Indicates the number of peers that this process is communicating with.  It correlates to the size of the peer array.

File: include/fi_shm.h (hunk 3 above)
Comment (shefty) [inline include/fi_shm.h]: That's the problem with learning C before 1999.  I remember compilers having trouble with this sort of array declaration.  A variable sized array is the intent.

File: include/fi_shm.h (hunk 5 above)
Comment (hppritcha) [inline include/fi_shm.h]: I think the design should be able to work with applications which don't know a priori what value to use for  ``` peer_count```.

Comment (shefty) [inline include/fi_shm.h]: While that would be nice, it greatly complicates the design and implementation.  And I would much rather have something available sooner, than defer it until dynamic support is available.  The current memory overhead of each peer is the size of a pointer, so setting peer_count = number of cores (or even a multiple of it) isn't a significant hit on memory, and may very well be smaller than dynamic support would require for a smaller number of peers.  (E.g. a fixed size peer_count = 100 with only 50 used may be less overhead than dynamic support for 50 peers.)

**Comment by shefty** [comment]
Updated based on comments.  Redefined the libfabric <-> libfabric protocol headers.

**Pull request closed without merging**